mater/lib/src/unixfs/mod.rs

```diff
@@ -41,7 +42,105 @@ enum TreeNode {
 }
 
 impl TreeNode {
-    fn encode(self) -> Result<((Cid, Bytes), LinkInfo), Error> {
+    fn encode_unixfs_leaf_node(chunk: &Bytes) -> Result<((Cid, Bytes), LinkInfo), Error> {
+        let chunk_len = chunk.len() as u64;
+
+        // Build UnixFS metadata
+        let unixfs_data = Data {
+            Type: mod_Data::DataType::File,
+            filesize: Some(chunk_len),
+            blocksizes: vec![chunk_len],
+            Data: Some(chunk.to_vec().into()),
+            hashType: None,
+            fanout: None,
+        };
+
+        // Encode UnixFS data and create DAG-PB node
+        let mut data_buf = Vec::new();
+        {
+            let mut w = quick_protobuf::Writer::new(&mut data_buf);
+            unixfs_data.write_message(&mut w)?;
+        }
+
+        let pb_node = PbNode {
+            links: vec![],
+            data: Some(data_buf.clone().into()),
+        };
+
+        let encoded = DagPbCodec::encode_to_vec(&pb_node)?;
+        let mh = generate_multihash::<Sha256, _>(&encoded);
+        let cid = Cid::new_v1(DAG_PB_CODE, mh);
+
+        let info = LinkInfo {
+            raw_data_length: chunk_len,
+            encoded_data_length: encoded.len() as u64,
+        };
+
+        Ok(((cid, encoded.into()), info))
+    }
+
+    fn encode_unixfs_stem_node(
+        children: Vec<(Cid, LinkInfo)>,
+    ) -> Result<((Cid, Bytes), LinkInfo), Error> {
+        // Process all children in a single pass, gathering totals and building links and blocksizes
+        let (total_raw_size, total_encoded_size, pb_links, blocksizes) = children.iter().fold(
+            (
+                0u64,
+                0u64,
+                Vec::with_capacity(children.len()),
+                Vec::with_capacity(children.len()),
+            ),
+            |(raw_sum, encoded_sum, mut links, mut sizes), (child_cid, link_info)| {
+                sizes.push(link_info.raw_data_length);
+                links.push(PbLink {
+                    cid: *child_cid,
+                    name: Some("".to_string()),
+                    size: Some(link_info.encoded_data_length),
+                });
+                (
+                    raw_sum + link_info.raw_data_length,
+                    encoded_sum + link_info.encoded_data_length,
+                    links,
+                    sizes,
+                )
+            },
+        );
```

Comment:
IMO this is much simpler and more readable:
```suggestion
        let mut total_raw_size: u64 = 0;
        let mut total_encoded_size: u64 = 0;
        let mut pb_links = Vec::with_capacity(children.len());
        let mut blocksizes = Vec::with_capacity(children.len());
        // Process all children in a single pass, gathering totals and building links and blocksizes
        for (cid, link_info) in children {
            pb_links.push(PbLink {
                cid,
                name: Some("".to_string()),
                size: Some(link_info.encoded_data_length),
            });
            blocksizes.push(link_info.raw_data_length);

            total_raw_size += link_info.raw_data_length;
            total_encoded_size += link_info.encoded_data_length;
        }
```